:sparkles: Apply PodSecurityStandard in Secure Cluster Class

chrischdi opened the pull request:
<!-- please add a icon to the title of this PR (see https://sigs.k8s.io/cluster-api/VERSIONING.md), and delete this line and similar ones -->
<!-- the icon will be either ⚠️ (:warning:, major or breaking changes), ✨ (:sparkles:, feature additions), 🐛 (:bug:, patch and bugfixes), 📖 (:book:, documentation or proposals), or 🌱 (:seedling:, minor or other) -->

**What this PR does / why we need it**:

Adds a [baseline pod security standard](https://kubernetes.io/docs/concepts/security/pod-security-standards/#baseline) which gets enforced at cluster level.


**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Resolves parts of #6329

Open for discussion for me:

Wether or not add it to the default ClusterClass or add a separate ClusterClass e.g. called `capi-quickstart-secure`.
A second ClusterClass would have the con of duplicating the normal quick-start ClusterClass with lots of duplicated content.

If adding the PSS to the default cluster class results into issues we could also work the other way around and add a "insecure" ClusterClass named `capi-quickstart-secure` fur users to fallback to (`clusterctl generate cluster capi-quickstart --flavor development-topology-insecure ...`).

TODO:
- [x] squash commits

killianmuldoon commented:
Awesome - this is working for me:
```bash
The Pod "busybox" is invalid: spec: Forbidden: pod updates may not change fields other than `spec.containers[*].image`, `spec.initContainers[*].image`, `spec.activeDeadlineSeconds` or `spec.tolerations` (only additions to existing tolerations)
  core.PodSpec{
        Volumes:        {{Name: "kube-api-access-p9jbd", VolumeSource: {Projected: &{Sources: {{ServiceAccountToken: &{ExpirationSeconds: 3607, Path: "token"}}, {ConfigMap: &{LocalObjectReference: {Name: "kube-root-ca.crt"}, Items: {{Key: "ca.crt", Path: "ca.crt"}}}}, {DownwardAPI: &{Items: {{Path: "namespace", FieldRef: &{APIVersion: "v1", FieldPath: "metadata.namespace"}}}}}}, DefaultMode: &420}}}},
        InitContainers: nil,
        Containers: []core.Container{
                {
                        ... // 16 identical fields
                        TerminationMessagePolicy: "File",
                        ImagePullPolicy:          "IfNotPresent",
-                       SecurityContext:          &core.SecurityContext{Privileged: &true},
+                       SecurityContext:          nil,
                        Stdin:                    false,
                        StdinOnce:                false,
                        TTY:                      false,
                },
        },
        EphemeralContainers: nil,
        RestartPolicy:       "Always",
        ... // 5 identical fields
        AutomountServiceAccountToken: nil,
        NodeName:                     "cloister-linux-workers-ppp8v-775fffb45b-kdp7k",
        SecurityContext: &core.PodSecurityContext{
-               HostNetwork: true,
+               HostNetwork: false,
                HostPID:     false,
                HostIPC:     false,
                ... // 11 identical fields
        },
        ImagePullSecrets: nil,
        Hostname:         "",
        ... // 15 identical fields
  }
```

We should run the full e2e suite on this before merging to make sure it doesn't break anything (as this modifies the ClusterClass that underpins all of our tests. I don't think there should be an issue, but just in case.

PushkarJ commented:
@chrischdi Thank you so much for getting around to opening this PR so quickly!!!!

PushkarJ commented:
> If adding the PSS to the default cluster class results into issues
What would be the definition of "no issues detected by this change?". If it is only about e2e test passing then looks like we have achieved that already!! 🎉 

> we could also work the other way around and add a "insecure" ClusterClass named capi-quickstart-insecure for users to fallback to (clusterctl generate cluster capi-quickstart --flavor development-topology-insecure ...)

Yeah that makes sense to me because it meets second and third guideline on secure defaults in cloud native apps mentioned here: https://github.com/cncf/tag-security/blob/main/security-whitepaper/secure-defaults-cloud-native-8.md

Next step for me sounds like figuring out how we can enable this across all CAPI cloud providers.

killianmuldoon commented:
I think we should be happy with adding this as is for now as the main use for this ClusterClass is e2e testing and a reference implementation.

/lgtm

chrischdi commented:
For the cloud providers there currently are no pre-generated/pre-defined cluster classes to use (I checked the latest releases of [CAPA](https://github.com/kubernetes-sigs/cluster-api-provider-aws/releases/tag/v1.4.0), [CAPZ](https://github.com/kubernetes-sigs/cluster-api-provider-azure/releases/tag/v1.2.1) and [CAPG](https://github.com/kubernetes-sigs/cluster-api-provider-gcp/releases/tag/v1.0.2)). 

From a clusterctl perspective: it is only consuming the released yaml files of the infrastructure providers (in case of this PR the docker infrastructure provider).
The docker infrastructure provider is the only one currently providing cluster class manifests.

One thing we could do to improve here would be to add a security guidelines section to the book [as proposed by fabrizzio here](https://github.com/kubernetes-sigs/cluster-api/issues/6329#issuecomment-1090076362).

fabriziopandini commented:
> Whether or not add it to the default ClusterClass or add a separate ClusterClass e.g. called capi-quickstart-secure.
A second ClusterClass would have the con of duplicating the normal quick-start ClusterClass with lots of duplicated content.

> If adding the PSS to the default cluster class results into issues we could also work the other way around and add a "insecure" ClusterClass named capi-quickstart-secure fur users to fallback to (clusterctl generate cluster capi-quickstart --flavor development-topology-insecure ...).

What if we add a variable, named `disable-PodSecurityStandard` defaulted to false (and thus optional), and we use this var to add/remove the changes in this PR via patches (instead of changing editing templates sources).

Pushing this a little bit further, we can make this variable a struct allowing users to define:
- Pod Security standard name (restricted, baseline, privileged)
- Applicable mode (enforce, warn, audit)
- Applicable version (usually default)
(this can be also a follow up iteration)

Also, what about extending the scope of this PR including the creation of a security guidelines section n the book (probably under tasks) with a short description to how to enable pod-security-standards + an inline version of the above variable/patches and a link to the CAPD example

chrischdi commented:
> > Whether or not add it to the default ClusterClass or add a separate ClusterClass e.g. called capi-quickstart-secure.
> > A second ClusterClass would have the con of duplicating the normal quick-start ClusterClass with lots of duplicated content.
> 
> > If adding the PSS to the default cluster class results into issues we could also work the other way around and add a "insecure" ClusterClass named capi-quickstart-secure fur users to fallback to (clusterctl generate cluster capi-quickstart --flavor development-topology-insecure ...).
> 
> What if we add a variable, named `enable-PodSecurityStandard` defaulted to true, and we use this var to add/remove the changes in this PR via patches (instead of changing editing templates sources).
> 
> Pushing this a little bit further, we can make this variable a struct allowing users to define:
> 
> * Pod Security standard name (restricted, baseline, privileged)
> * Applicable mode (enforce, warn, audit)
> * Applicable version (usually default)
>   (this can be also a follow up iteration)
> 
> Also, what about extending the scope of this PR including the creation of a security guidelines section n the book (probably under tasks) with a short description to how to enable pod-security-standards + an inline version of the above variable/patches and a link to the CAPD example

I did some refactoring of the PR :-)

* I added a todo at the top-level comment so I don't forget to squash commits
* Refactored the configuration to use patches via cluster class instead which allows configuring the levels and disabling the patches (keeping the default enabled)
* book/quick-start: added a short part to mention how PSS could be disabled for the docker quickstart
* book/writing cluster class: added a Limitation section to note the unability to append to empty slices via json patches
* book/security-guidelines: added a section which as of now only contains adding pod security policy to a cluster class

PushkarJ commented on docs/book/src/security-guidelines/index.md:
```diff
@@ -0,0 +1,6 @@
+# Security Guidelines
+
+This section provides security guidelines useful to provision clusters which are
+secure by default to follow the [secure defaults guidelines for cloud native apps].
```
```suggestion
_secure by default_ to follow the [secure defaults guidelines for cloud native apps].
```

PushkarJ commented on docs/book/src/security-guidelines/pod-security-standards.md:
```diff
@@ -0,0 +1,211 @@
+# Pod Security Standards
+
+Pod Security Admission allows defining [Pod Security Standards] for creation of pods at the cluster level.
```
```suggestion
Pod Security Admission allows applying [Pod Security Standards] during creation of pods at the cluster level.
```

PushkarJ commented on docs/book/src/tasks/experimental-features/cluster-class/write-clusterclass.md:
```diff
@@ -676,6 +677,101 @@ individually.
+* the slice was defined with `omitempty`
+* the slice is currently does not exist
+
+A workaround in this particular case is to create the slice in the patch instead of adding to the not existing one.
```
should this be`... instead of adding it to the non-existing one` ?

PushkarJ commented on docs/book/src/tasks/experimental-features/cluster-class/write-clusterclass.md:
```diff
@@ -676,6 +677,101 @@ individually.
+* the patch is a JSON patch
+* the patch tries to `add` a value to a **slice**
+* the slice was defined with `omitempty`
+* the slice is currently does not exist
```
```suggestion
* the slice currently does not exist
```

PushkarJ commented on docs/book/src/security-guidelines/pod-security-standards.md:
````diff
@@ -0,0 +1,211 @@
+            pattern: "privileged|baseline|restricted"
+  ...
+```
+
````
Maybe worth adding a couple of notes:
- The version field in Pod Security Admission Config defaults to `latest` 
- The `kube-system` namespace is exempt from Pod Security Standards enforcement, because it runs control-plane pods that need higher privileges

PushkarJ commented on docs/book/src/security-guidelines/pod-security-standards.md:
```diff
@@ -0,0 +1,211 @@
+
+### Create a secure Cluster using the ClusterClass
+
+After adding the variables and patches the Pod Security Standard would be added by default.
```
```suggestion
After adding the variables and patches the Pod Security Standards would be applied by default.
```

PushkarJ reviewed (commented):
Few suggestions on docs for clarity. But overall looks good!!

PushkarJ commented:
Would it make sense to add the content from #6152 in to the security guidelines section created in this PR?

PushkarJ commented:
/retitle Apply PodSecurityStandard in Secure Cluster Class

(Since the scope of the PR has changed thought this might be a better reflection of the work being done. Please revert/edit as needed)

killianmuldoon commented:
/retitle :sparkles: Apply PodSecurityStandard in Secure Cluster Class

Just adding a prefix to make this work for our release note generation process :slightly_smiling_face:

chrischdi commented:
Thanks for the detailed review 👍 

> Would it make sense to add the content from #6152 in to the security guidelines section created in this PR?
> 
> Edit: To be clear, Not suggesting to add it in this PR but in follow up PRs :)

Should be fine to add the content of #6152 in a separate PR.

fabriziopandini commented on docs/book/src/security-guidelines/index.md:
```diff
@@ -0,0 +1,6 @@
+# Security Guidelines
```
nit, not blocking 
I think we can drop `-guidelines` from the folder name (for brevity)

chrischdi replied:
Valid point :-) renamed the directory

fabriziopandini commented on docs/book/src/security-guidelines/pod-security-standards.md:
```diff
@@ -0,0 +1,214 @@
+      value: 
+        enabled: true
+        enforce: "restricted"
+        audit: "restricted"
+        warn: "restricted"
```
nit: Should we drop this given that `restricted` is the default

fabriziopandini commented on docs/book/src/tasks/experimental-features/cluster-class/write-clusterclass.md:
```diff
@@ -676,6 +677,101 @@ individually.
 
 </aside>
 
+## Limitations of ClusterClass with patches
```
```suggestion
## JSON patches tips
```

fabriziopandini commented on docs/book/src/tasks/experimental-features/cluster-class/write-clusterclass.md:
```diff
@@ -676,6 +677,101 @@ individually.
 
+## Limitations of ClusterClass with patches
+
+JSON patches are limited by the specification [RFC6902] in not being able to add values to not-existing targets.
```
```suggestion
JSON patches specification [RFC6902] requires that the target of
add operation must exist.
```

fabriziopandini commented on docs/book/src/tasks/experimental-features/cluster-class/write-clusterclass.md:
```diff
@@ -676,6 +677,101 @@ individually.
+
+JSON patches are limited by the specification [RFC6902] in not being able to add values to not-existing targets.
+
+The topology controller which applies the patches will return an error if a patch specified in 
+a ClusterClass matches all of the following conditions:
```
```suggestion
As a consequence ClusterClass authors should pay special attention when the following conditions apply in order to prevent errors when a patch is applied:
```

fabriziopandini reviewed (commented):
Last round of nits but really looking forward to get this merged

chrischdi commented:
Addressed comments and rebased for squashing :-)

chrischdi commented:
cert-manager failed. Should be unrelated

/test pull-cluster-api-e2e-main

fabriziopandini commented:
/lgtm

PushkarJ commented:
/lgtm

(Thank you so much for working on this @chrischdi !)

fabriziopandini commented:
/test pull-cluster-api-test-full-main

fabriziopandini commented:
/test pull-cluster-api-e2e-full-main

sbueringer commented on docs/book/src/tasks/experimental-features/cluster-class/write-clusterclass.md:
```diff
@@ -676,6 +677,102 @@ individually.
+As a consequence ClusterClass authors should pay special attention when the following
+conditions apply in order to prevent errors when a patch is applied:
+
+* the patch is a JSON patch
```
nit: I think this first point might be confusing as today we can only write JSON patches in the ClusterClass.

Future external Patches are another topic, but this document focuses a lot on writing inline paches

chrischdi replied:
The whole paragraph now is about JSON patches so this point should not be necessary. I removed it

sbueringer commented on docs/book/src/tasks/experimental-features/cluster-class/write-clusterclass.md:
```diff
@@ -676,6 +677,102 @@ individually.
+conditions apply in order to prevent errors when a patch is applied:
+
+* the patch is a JSON patch
+* the patch tries to `add` a value to a **slice**
```
nit: we're mixing Go and JSON here a bit

Up to you but maybe there's a way to make it clearer when we are talking about the slice in an API type and when we are talking about an array in JSON

But maybe it's easier for users if we don't highlight that difference

chrischdi replied:
I agree, I changed to using the RFC wording `array` and added at this occurence: `(which is a **slice** in the corresponding go struct)`

sbueringer commented on docs/book/src/tasks/experimental-features/cluster-class/write-clusterclass.md:
```diff
@@ -676,6 +677,102 @@ individually.
+{{#/tab }}
+{{#tab Create slice}}
+
+This patch would **overwrite** an existing slice at `.spec.template.spec.files`.
```
I think it would fail because the JSON patch lib would detect that it can't add an already existing array (only with op: replace it would overwrite it)

chrischdi replied:
The library used in clusterclass for JSON patching would overwrite it.

```
apiVersion: bootstrap.cluster.x-k8s.io/v1beta1
kind: KubeadmConfigTemplate
spec:
  templates:
    spec:
      files:
      - content: Old
        path: /some
```

And the patch below it would result in
```
apiVersion: bootstrap.cluster.x-k8s.io/v1beta1
kind: KubeadmConfigTemplate
spec:
  templates:
    spec:
      files:
      - content: Some content.
        path: /some/file
```

I will verify that again tomorrow.

I agree with you, normally it should be invalid, e.g. https://json-schema-validator.herokuapp.com/jsonpatch.jsp would throw an error.

sbueringer replied:
Wow, I thought I remembered having that error. But it's absolutely possible that I'm wrong.

Just one fyi in case that's somehow relevant. I think during some version bump the jsonpatch lib (https://github.com/evanphx/json-patch) got more strict. I think it was from 4 to 5.

chrischdi replied:
I'm pretty sure I had been on main at the time of testing. But I will verify that.

chrischdi replied:
Ah, its the library which actually allows it https://go.dev/play/p/bT8wcbh7BSQ

sbueringer replied:
Okay then you're right :). Maybe it was the old version which was more strict

sbueringer reviewed (commented):
Very nice

I assume the current state is manually tested.

It would be really nice to see a follow-up PR to include those changes in our e2e test quickstart ClusterClass to:
* find out if this is compatible with our e2e tests
* to keep the diff between tested templates/ClusterClass and released template as small as possible

chrischdi commented:
Thanks for the detailed review. I applied changes according the following patch:
```diff
--- a/docs/book/src/security/pod-security-standards.md
+++ b/docs/book/src/security/pod-security-standards.md
@@ -3,7 +3,7 @@
 Pod Security Admission allows applying [Pod Security Standards] during creation of pods at the cluster level.

 The flavor `development-topology` for the docker provider used in [Quick Start](../user/quick-start.md) already includes a basic Pod Security Standard configuration.
-It is using variables and patches via ClusterClass to inject the configuration.
+It is using ClusterClass variables and patches to inject the configuration.

 ## Adding a basic Pod Security Standards configuration to a ClusterClass

@@ -51,7 +51,7 @@ spec:

 The following snippet contains the patch to be added to the ClusterClass.

-Due to [limitations of ClusterClass with patches](../tasks/experimental-features/cluster-class/write-clusterclass.md#limitations-of-clusterclass-with-patches) there are two versions for this patch.
+Due to [limitations of ClusterClass with patches](../tasks/experimental-features/cluster-class/write-clusterclass.md#json-patches-tips--tricks) there are two versions for this patch.

 {{#tabs name:"tab-configuration-patches" tabs:"Add to existing slice,Create slice"}}
 {{#tab Append}}
@@ -76,7 +76,7 @@ spec:
         matchResources:
           controlPlane: true
       jsonPatches:
-       - op: add
+      - op: add
         path: "/spec/template/spec/kubeadmConfigSpec/clusterConfiguration/apiServer/extraArgs"
         value:
           admission-control-config-file: "/etc/kubernetes/kube-apiserver-admission-pss.yaml"
@@ -133,7 +133,7 @@ kind: ClusterClass
 spec:
   ...
   patches:
-  - name: admissionPodSecurityPolicy
+  - name: podSecurityStandard
     description: "Adds an admission configuration for PodSecurity to the kube-apiserver."
     definitions:
     - selector:
@@ -190,7 +190,7 @@ spec:

 ### Create a secure Cluster using the ClusterClass

-After adding the variables and patdocs/book/src/tasks/experimental-features/cluster-class/write-clusterclass.mdches the Pod Security Standards would be applied by default.
+After adding the variables and patches the Pod Security Standards would be applied by default.
 It is also possible to disable this patch or configure different levels for the configuration
 using variables.

diff --git a/docs/book/src/tasks/experimental-features/cluster-class/write-clusterclass.md b/docs/book/src/tasks/experimental-features/cluster-class/write-clusterclass.md
index 32b6a1dc6..4d62bf6c7 100644
--- a/docs/book/src/tasks/experimental-features/cluster-class/write-clusterclass.md
+++ b/docs/book/src/tasks/experimental-features/cluster-class/write-clusterclass.md
@@ -685,12 +685,11 @@ add operation must exist.
 As a consequence ClusterClass authors should pay special attention when the following
 conditions apply in order to prevent errors when a patch is applied:

-* the patch is a JSON patch
-* the patch tries to `add` a value to a **slice**
+* the patch tries to `add` a value to an **array** (which is a **slice** in the corresponding go struct)
 * the slice was defined with `omitempty`
 * the slice currently does not exist

-A workaround in this particular case is to create the slice in the patch instead of adding it to the non-existing one.
+A workaround in this particular case is to create the array in the patch instead of adding to the non-existing one.
 When creating the slice, existing values would be overwritten so this should only be used when it does not exist.

 The following example shows both cases to consider while writing a patch for adding a value to a slice.
```

I will verify again the last open point tomorrow.

sbueringer commented:
Check, looks good so far

sbueringer commented:
/lgtm

fabriziopandini reviewed (commented):
/lgtm
/approve
